### PR TITLE
feat(pallets/subspace): add faucet function

### PR DESCRIPTION
**Pull request by aripiprazole**

### CODE REVIEW

File: pallets/subspace/src/lib.rs [hunk 1]
```diff
@@ -1067,6 +1080,23 @@ pub mod pallet {
...
+        #[pallet::weight((Weight::from_parts(91_000_000, 0)
+		.saturating_add(T::DbWeight::get().reads(27))
+		.saturating_add(T::DbWeight::get().writes(22)), DispatchClass::Normal, Pays::No))]
+        pub fn faucet(
+            origin: OriginFor<T>,
+            block_number: u64,
+            nonce: u64,
+            work: Vec<u8>,
+        ) -> DispatchResult {
+            if cfg!(test) {
```
Comment (Supremesource) [inline pallets/subspace/src/lib.rs:1093]: rename this to testnet

File: pallets/subspace/src/registration.rs [hunk 2]
```diff
@@ -257,6 +259,121 @@ impl<T: Config> Pallet<T> {
...
+        // --- 6. Deposit successful event.
+        log::info!("Faucet(coldkey: {coldkey:?} amount: {balance_to_add:?})",);
+        Self::deposit_event(Event::Faucet(coldkey, balance_to_add));
+
+        // --- 7. Ok and done.
+        Ok(())
+    }
+
+    #[allow(clippy::indexing_slicing)]
+    pub fn hash_block_and_hotkey(block_hash_bytes: &[u8; 32], hotkey: &T::AccountId) -> H256 {
```
Comment (Supremesource) [inline pallets/subspace/src/registration.rs:310]: rename everything from hotkey to just `key` commune does not have hotkeys

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -366,6 +366,14 @@ impl pallet_subspace::Config for Runtime {
...
 }
 
+impl BlockNumberProvider for Runtime {
+    type BlockNumber = BlockNumber;
+
+    fn current_block_number() -> Self::BlockNumber {
+        System::block_number()
+    }
+}
+
```
Comment (saiintbrisson) [inline runtime/src/lib.rs:376]: @aripiprazole I don't think we need this anymore?

File: pallets/subspace/tests/mock.rs [hunk 4]
```diff
@@ -110,6 +110,14 @@ impl pallet_subspace::Config for Test {
...
 }
 
+impl BlockNumberProvider for Test {
+    type BlockNumber = u64;
+
+    fn current_block_number() -> Self::BlockNumber {
+        System::block_number()
+    }
+}
+
```
Comment (saiintbrisson) [inline pallets/subspace/tests/mock.rs:120]: Or this

**Review by saiintbrisson — CHANGES_REQUESTED** [review]

**Review by saiintbrisson — APPROVED** [review]

**Pull request merged**